internal/engine/netx/errorx/errorx.go

```diff
@@ -244,50 +280,83 @@ func toFailureString(err error) string {
 	if strings.HasSuffix(s, "TLS handshake timeout") {
 		return FailureGenericTimeoutError
 	}
-	if strings.HasSuffix(s, "no such host") {
+	if strings.HasSuffix(err.Error(), "no such host") {
 		// This is dns_lookup_error in MK but such error is used as a
 		// generic "hey, the lookup failed" error. Instead, this error
 		// that we return here is significantly more specific.
 		return FailureDNSNXDOMAINError
 	}
+	formatted := fmt.Sprintf("unknown_failure: %s", s)
+	return Scrub(formatted) // scrub IP addresses in the error
+}
 
-	// TODO(kelmenhorst): see whether it is possible to match errors
-	// from qtls rather than strings for TLS errors below.
-	//
-	// TODO(kelmenhorst): make sure we have tests for all errors. Also,
-	// how to ensure we are robust to changes in other libs?
-	//
-	// special QUIC errors
-	matched, err := regexp.MatchString(`.*x509: certificate is valid for.*not.*`, s)
-	if matched {
-		return FailureSSLInvalidHostname
+func ClassifyQUICFailure(err error) string {
+	var versionNegotiation *quic.VersionNegotiationError
+	var statelessReset *quic.StatelessResetError
+	var handshakeTimeout *quic.HandshakeTimeoutError
+	var idleTimeout *quic.IdleTimeoutError
+	var transportError *quic.TransportError
+
+	if errors.As(err, &versionNegotiation) {
+		return FailureNoCompatibleQUICVersion
 	}
-	if strings.HasSuffix(s, "x509: certificate signed by unknown authority") {
-		return FailureSSLUnknownAuthority
+	if errors.As(err, &statelessReset) {
+		return FailureConnectionReset
 	}
-	certInvalidErrors := []string{"x509: certificate is not authorized to sign other certificates", "x509: certificate has expired or is not yet valid:", "x509: a root or intermediate certificate is not authorized to sign for this name:", "x509: a root or intermediate certificate is not authorized for an extended key usage:", "x509: too many intermediates for path length constraint", "x509: certificate specifies an incompatible key usage", "x509: issuer name does not match subject from issuing certificate", "x509: issuer has name constraints but leaf doesn't have a SAN extension", "x509: issuer has name constraints but leaf contains unknown or unconstrained name:"}
-	for _, errstr := range certInvalidErrors {
-		if strings.Contains(s, errstr) {
+	if errors.As(err, &handshakeTimeout) {
+		return FailureGenericTimeoutError
+	}
+	if errors.As(err, &idleTimeout) {
+		return FailureGenericTimeoutError
+	}
+	if errors.As(err, &transportError) {
+		if transportError.ErrorCode == quic.ConnectionRefused {
+			return FailureConnectionRefused
+		}
+		// checkout alert.go in the qtls library
+		// (the alert constants are not exported, so this is not robust to changes in qtls)
+		errCode := uint8(transportError.ErrorCode)
+		if isCertificateError(errCode) {
 			return FailureSSLInvalidCertificate
 		}
+		if errCode == TLSDecryptError || errCode == TLSHandshakeFailure {
+			return FailureSSLHandshake
```

Comment:
Do we perhaps want to treat these two as separate errors? What is the rationale of coalescing them?

Reply:
(I see this is a new error that we were not previously reporting, right? So a good question would probably be if we could not take advantage of existing errors? Is this a new error condition? If so, could we take advantage of this error condition and, instead, perhaps _also_ use this error for non-quic handshakes?)

Reply:
the rationale was that both errors are caused by a failed or corrupted parameter negotiation during the TLS handshake

Reply:
could you add a comment mentioning that please?